### ISSUE TITLE
Incompatibility with Rstudio internal browser

### ISSUE THREAD

**Issue by juba** [d3/d3-timer#16]
Hi,

I'm using [RStudio](https://www.rstudio.com/), an R language IDE, for developing an [htmlwidget](http://www.htmlwidgets.org/) with d3. I'm trying to port my code to d3 v4, and I'm facing a curious problem: transitions don't work anymore. They work fine with d3 v3 in RStudio or with d3 v4 in a browser, but not with d3 v4 in RStudio.

After a bit of digging, it seems that the problem comes from the toolkit used by RStudio (GWT 2.7.0 as far as I can tell, but I don't know what kind and version of browser it is. The user agent displayed in the "about" dialog is currently _"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/538.1 (KHTML, like Gecko) RStudio Safari/538.1 Qt/5.4.0"_).

There seems to be an inconsistency in the results returned by `performance.now` and `requestAnimationFrame` : 

```js
> performance.now()
26015828.386478003
> requestAnimationFrame(function(timestamp) {console.log(timestamp)})
1
1472136450436.1724
```

If I understand correctly, It seems that `performance.now()` returns an high precision timestamp, whereas `requestAnimationFrame` gives a number of milliseconds since epoch.

Due to this inconsistency, timers and transitions created by d3 don't work at all.

I've found a quick and dirty workaround inspired by [this link](https://developers.google.com/web/updates/2012/05/requestAnimationFrame-API-now-with-sub-millisecond-precision#feature-detection). Adding the following code in `d3-timer.js` just before the `now()` function definition seems to work :

```js
  if (typeof performance === "object" && clock == performance && typeof requestAnimationFrame === "function") {
    requestAnimationFrame(function(timestamp){
      if (timestamp > 1e12) {
        clock = Date;
        setFrame = function(f) { requestAnimationFrame(function() { f(clock.now()); }); }
      }
    })
  }
```

I could generate a pull request for this if you wish, but I'm really not sure the code quality is good enough, neither that this kind of workaround would really have a place in `d3` code. Or there may be a better solution.

Sorry for this quite long issue, and thanks a lot for this awesome library and the no less awesome documentation and examples !

**Comment by juba** [comment]
Thanks for the detailed feedback.

As far as I can tell, the only callback function called by `setFrame` and which uses the timestamp passed as argument is `wake`. So it seems that adding this at the start of the `wake` definition should do the trick :

```js
  if (time > 1e12 && typeof requestAnimationFrame === "function") {
      time = clock.now();
      setFrame = function(f) { requestAnimationFrame(function() { f(clock.now()); }); }
  }
```

The `setFrame` redefinition should only be called once, but the test on `time > 1e12` would be run at each `wake` call, and I don't really measure if it would lead to a performance loss...

Sorry if this is not the solution you were suggesting, as I'm far from a javascript expert.

**Comment by mbostock** [comment]
I thought about this some more, and I think it will be simplest to just always call performance.now (if available, and otherwise Date.now) rather than rely on the timestamp passed to the requestAnimationFrame callback.

**Comment by mbostock** [comment]
Fixed in @e8ac3d2.

**Comment by juba** [comment]
Just tested it on RStudio, timers and transitions are working fine again. Thanks a lot !
